### ISSUE TITLE
returned object from useModal hook have "unstable" identity

### ISSUE THREAD

**Issue by mrudowski** [eBay/nice-modal-react#58]
Hi, so nice library!

and this fix for 
https://github.com/eBay/nice-modal-react/issues/53

just in time (I've just updated to 1.2.4)
Thank you!

But could we do the same for **whole object returned** from useModal?
```
const userModal = useModal(UserInfoModal);
```

to make `userModal` referential equality all the time?
so if I test it here like this:

```
const userModal = useModal(UserInfoModal);

React.useEffect(() => {
    console.log('endless show...');
    userModal.show();
}, [userModal]) // <- referential identity will change every time and we have endless loop here
```

**Comment by mrudowski** [comment]
after reading this discussion at https://github.com/tannerlinsley/react-query/issues/1858 
I have to say it's not so simple... and maybe not quite in the right direction

after `react-hook-form` which return "stable" object from `useForm` hook I assumed that this is the way.
But now I dunno

**Comment by supnate** [comment]
Right, this is something we can enhance but not that simple. For now, you can destruct the methods `show`, `hide` from `userModal` as dependencies to avoid loops.

**Comment by janczizikow** [comment]
Just adding my two cents 😄  

There are many ways to avoid the infinite re-rendering, one of which was already mentioned (destructuring).

Other options:

Pass reference to a callback in useEffect (without destructuring):

```tsx
const userModal = useModal(UserInfoModal);

React.useEffect(() => {
    console.log('endless show...');
    userModal.show();
}, [userModal.show]) // <-- pass userModal.show
```


Pass whole object, but call the callback conditionally:

```tsx
const userModal = useModal(UserInfoModal);

React.useEffect(() => {
    if (!userModal.visible) { // <-- just as an example
      console.log('endless show...');
      userModal.show();
    }
}, [userModal])
```

As for memoizing the return value of useModal - as you mentioned it's probably not so easy. One way that I could think of could be some kind shallow equality checks of `userModal.args` either in the reducer in the `useModal` hook 🤔 But I think that will add more complexity to the lib.

**Comment by mrudowski** [comment]
> @mrudowski did you manage to make react-hook-form work with modals in react-nice-modal?

I use it all the time without any problems. I was only wondering about returning object from any hook and the right approach to it in react/hook world.

**Comment by supnate** [comment]
should be resolved in `v1.2.11`.

**Comment by nik32** [comment]
> should be resolved in `v1.2.11`.

@supnate does this mean that from `v1.2.11`... we can use the `userModal` inside of the effect dependencies array without having to worry about infinite re-renders? 
OR we would still need to destruct the methods `show`, `hide` from `userModal` as dependencies to avoid loops.

**Comment by xeinebiu** [comment]
@janczizikow 

`
const userModal = useModal(UserInfoModal);

React.useEffect(() => {
    console.log('endless show...');
    userModal.show();
}, [userModal.show]) // <-- pass userModal.show
`

This wont work with prettier nor eslint. They do not allow to pass functions due to "this" binding. 

Even destructuring the `show` function, it stills ends up on infinite loop. I think this library should use internally `useCallbacks` so it returns the same functions when nothing really changes, otherwise on this Issue only workarounds are proposed.

Version `1.2.23` and the loop still exists.
